styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -674,17 +687,63 @@ private String readStatus(Pod pod) {
 
     private static final int RECONNECT_DELAY_SECONDS = 1;
 
+    private final ConcurrentMap<String, WorkflowInstance> podUpdates = new ConcurrentHashMap<>();
+
     @Override
     public void eventReceived(Action action, Pod pod) {
+
+      // NOTE: In order to be able to keep up with the stream of events from k8s, this method should
+      //       not perform any expensive processing or blocking IO.
+
       if (pod == null) {
         return;
       }
 
       logEvent(action, pod, pod.getMetadata().getResourceVersion(), false);
 
-      readPodWorkflowInstance(pod)
-          .flatMap(workflowInstance -> lookupPodRunState(pod, workflowInstance))
-          .ifPresent(runState -> emitPodEvents(action, pod, runState));
+      // Ignore pod deletions
+      if (action == Action.DELETED) {
+        return;
+      }
+
+      // Ignore non-styx pods
+      final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
+      if (!workflowInstance.isPresent()) {
+        return;
+      }
+
+      // Flag this pod for later processing. Note that instead of storing the received pod status here,
+      // a new status is fetched later, in order to avoid observing pod statuses out-of-order.
+      podUpdates.put(pod.getMetadata().getName(), workflowInstance.get());
+    }
+
+    void processPodUpdates() {
+      // Get a batch of pods to process
+      final Set<String> podNames = ImmutableSet.copyOf(podUpdates.keySet());
+
+      // Process the batch in parallel
+      podNames.stream()
+          .map(podName -> {
+            // Remove from change set before processing in order to not lose updates
+            final WorkflowInstance instance = podUpdates.remove(podName);
+            return CompletableFuture.runAsync(() -> processPodUpdate(podName, instance), eventExecutor);
+          })
+          .collect(Collectors.toList())
+          .forEach(CompletableFuture::join);
+    }
+
+    private void processPodUpdate(String podName, WorkflowInstance instance) {
+      final Pod pod = client.pods().withName(podName).get();
```

Comment:
This will result lots of additional pod readings so we should test this thoroughly in staging.